canonical/microcloud-cluster-manager
# preparation for network listener update in microcluster [WD-13102]

## Done

- [x] setup control listener to listen on different address than core listener
- [x] moved `POST /1.0/sites` endpoint to control listener
- [x] initialise control listener in `PostBootstrap` and `PostJoin` hooks. In addition, the initialisation will add a entry in the `manager_member_configs` table for each member with `https_address` populated. This also means we can remove logic populating `https_address` in the `generate_configs.sh` script.
- [x] setup global site manager state to hold `microcluster.Microcluster`. We pass this state into some endpoints so we can access client functions e.g. for updating network listener addresses
- [x] updated `PATCH /1.0/member/{name}/config` endpoint so that we will forward the request to the correct member if `https_address` is present in the payload. We will still populate the data inside dqlite so that read operations will be much easier to handle (trading off with some data redundancy)
- [x] to ensure config data is synced between daemon local config and dqlite, added a revert function to rollback a network listener update if the corresponding db transaction fails

## Context
 - This PR is created in preparation for the microcluster [update](https://github.com/canonical/microcluster/pull/174) that allows modification to network listeners.
 - **Currently this PR won't work since dependencies are not yet updated.**

## Follow up task
 - cache control listener and external addresses in the daemon state. This will improve efficiency for status updates.

## Testing
 - setup cluster with multiple members
 - update `https_address` config for individual members with `lxd-site-mgr config set <member> https_address=<address>`. 
 - make sure it works exactly as before.
 - @edlerd it would be great if you can test the join process since `POST /1.0/sites` is on a different listener now. I've tested it locally already but I think it would be good if you can check it too.
